drivers: regulator: list supported levels

etienne-lms opened the pull request:
Adds regulator API function regulator_supported_voltages() to get the list of the voltage levels supported by the regulator.

Voltage level array is either an array of increasing ordered levels, in microvolt, or is a triplet [min, max, step] for linear step incremental levels.

This P-R is extracted from RFC https://github.com/OP-TEE/optee_os/pull/6301.

<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         https://optee.readthedocs.io/en/latest/general/contribute.html

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" in the contribution guidelines.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc. More details
       about this can also be found on the link provided above.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

jforissier commented on core/include/drivers/regulator.h:
```diff
@@ -76,13 +80,19 @@ struct regulator {
+ * @supported_voltages: Get an array of supported levels in microvolt (uV)
+ *      When successful, @count is the number of cells in @levels_us
+ *      or @count is 0 when @levels_us is a 3 cells array [min, max, step],
+ *      in microvolt, for linear incremental level lists.
```
I am not very comfortable with `count == 0` for a 3-cell array, and the fact that the array is not constant format nor self-describing. Wouldn't it be a bit easier to define a struct?
```C
/* Defines the format of struct voltages::entries */
enum voltage_type {
        VOLTAGE_TYPE_FULL_LIST, /* extensive list in uV */
        VOLTAGE_TYPE_INCREMENT  /* min, max, increment (in uV) */
};
struct voltages {
        enum voltage_type type;
        size_t num_levels;
        /* Dynamic part: num_levels entries */
        /* int entries[] */
};
```

etienne-lms replied:
ok, makes sense. I'll update.
Thanks for the proposal.

jforissier commented on core/include/drivers/regulator.h:
```diff
@@ -39,6 +39,37 @@ struct regu_dt_desc {
+ *
+ * If regulator_voltages::type is VOLTAGE_TYPE_FULL_LIST, then
+ * regulator_voltages@entries stores regulator_voltages::num_levels cells,
+ * listing supported voltage levels in uV from lowestto highest value.
```
Missing space `lowest to`

jforissier reviewed (commented):
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

etienne-lms commented:
I addressed the _lowestto_ typo, squashed the fixup commits and applied your review tag.
Thanks.

make check-rust failed with `No space left on device (os error 28)`.

jforissier commented:
> I addressed the _lowestto_ typo, squashed the fixup commits and applied your review tag. Thanks.

Thanks!

> make check-rust failed with `No space left on device (os error 28)`.

I'll see if I can find a way to save some space when running the Rust job.
